zebra-consensus/src/block/check.rs

```diff
@@ -25,3 +30,40 @@ pub fn is_coinbase_first(block: &Block) -> Result<(), Error> {
     }
     Ok(())
 }
+
+/// [3.9]: https://zips.z.cash/protocol/protocol.pdf#subsidyconcepts
+pub fn is_subsidy_correct(block: &Block) -> Result<(), Error> {
+    let height = block.coinbase_height().unwrap();
+
+    let coinbase = block.transactions.get(0).ok_or("no coinbase transaction")?;
+    let outputs = coinbase.outputs();
+
+    // Todo: we need the network here.
+    let network = Network::Mainnet;
+
+    let canopy_height = Canopy.activation_height(network).ok_or("no canopy")?;
+    if height >= canopy_height {
+        // dont validate canopy yet
+        return Ok(());
+    }
+
+    // validate founders reward and miner subsidy
+    if height > block::Height(0) && height <= block::Height(Params::LAST_FOUNDER_REWARD_HEIGHT) {
+        let block_subsidy = subsidies::block_subsidy(height, Network::Mainnet);
+        let miner_subsidy = subsidies::miner_subsidy(height, Network::Mainnet);
+        let mut valid_founders: bool = false;
+        let mut valid_miner: bool = false;
+        for o in outputs {
+            let value: i64 = o.value.try_into().unwrap();
+            if value == block_subsidy as i64 / 5 {
+                valid_founders = true;
+            } else if value == miner_subsidy as i64 {
+                valid_miner = true;
+            }
+        }
+        if valid_founders && valid_miner {
+            return Ok(());
+        }
+    }
```

Comment:
I'd switch this to use iterators

```suggestion
    if height > block::Height(0) && height <= block::Height(Params::LAST_FOUNDER_REWARD_HEIGHT) {
        let block_subsidy = subsidies::block_subsidy(height, Network::Mainnet) as i64;
        let miner_subsidy = subsidies::miner_subsidy(height, Network::Mainnet) as i64;
        let values = || {
            outputs
                .iter()
                .map(|o| o.value.try_into().expect("value will be a valid amount"))
        };

        if values().any(|value: i64| value == block_subsidy / 5 && value != miner_subsidy)
            && values().any(|value| value == miner_subsidy && value != block_subsidy / 5)
        {
            return Ok(());
        }
    }
```

Reply:
This is excellent, i will try to learn from it going further. Applied as it is at https://github.com/ZcashFoundation/zebra/pull/1051/commits/5636f56482df9b2320bf5bce0d0704c5c61f279f